In [1]:
import psycopg2 as pg
import pandas as pd

conn = pg.connect(database="latihan",
                 user="postgres",
                 password="ashalandorei",
                 host="127.0.0.1",
                 port="5432")

def run(query):
    df = pd.read_sql(query,conn)
    return df

# 1. Prepare a list of offices sorted by country, state, city.

In [2]:
run("""
SELECT COUNTRY, STATE, CITY
FROM _OFFICES
""")

country       state           city
0        USA          CA  San Francisco
1        USA          MA         Boston
2        USA          NY            NYC
3     France                      Paris
4      Japan  Chiyoda-Ku          Tokyo
5  Australia                     Sydney
6         UK                     London

# 2. How many employees are there in the company?

In [3]:
run("""
SELECT *
FROM _employees
""")

employeenumber   lastname firstname extension  \
0             1002     Murphy     Diane     x5800   
1             1056  Patterson      Mary     x4611   
2             1076   Firrelli      Jeff     x9273   
3             1088  Patterson   William     x4871   
4             1102     Bondur    Gerard     x5408   
5             1143        Bow   Anthony     x5428   
6             1165   Jennings    Leslie     x3291   
7             1166   Thompson    Leslie     x4065   
8             1188   Firrelli     Julie     x2173   
9             1216  Patterson     Steve     x4334   
10            1286      Tseng  Foon Yue     x2248   
11            1323     Vanauf    George     x4102   
12            1337     Bondur      Loui     x6493   
13            1370  Hernandez    Gerard     x2028   
14            1401   Castillo    Pamela     x2759   
15            1501       Bott     Larry     x2311   
16            1504      Jones     Barry      x102   
17            1611     Fixter      Andy      x101   
18            1612      Marsh     Peter      x102   
19            1619       King       Tom      x103   
20            1621      Nishi      Mami      x101   
21            1625       Kato   Yoshimi      x102   
22            1702     Gerard    Martin     x2312   

                              email  officecode reportsto  \
0      dmurphy@classicmodelcars.com           1             
1    mpatterso@classicmodelcars.com           1      1002   
2    jfirrelli@classicmodelcars.com           1      1002   
3   wpatterson@classicmodelcars.com           6      1056   
4      gbondur@classicmodelcars.com           4      1056   
5         abow@classicmodelcars.com           1      1056   
6    ljennings@classicmodelcars.com           1      1143   
7    lthompson@classicmodelcars.com           1      1143   
8    jfirrelli@classicmodelcars.com           2      1143   
9   spatterson@classicmodelcars.com           2      1143   
10      ftseng@classicmodelcars.com           3      1143   
11     gvanauf@classicmodelcars.com           3      1143   
12     lbondur@classicmodelcars.com           4      1102   
13   ghernande@classicmodelcars.com           4      1102   
14   pcastillo@classicmodelcars.com           4      1102   
15       lbott@classicmodelcars.com           7      1102   
16      bjones@classicmodelcars.com           7      1102   
17     afixter@classicmodelcars.com           6      1088   
18      pmarsh@classicmodelcars.com           6      1088   
19       tking@classicmodelcars.com           6      1088   
20      mnishi@classicmodelcars.com           5      1056   
21       ykato@classicmodelcars.com           5      1621   
22     mgerard@classicmodelcars.com           4      1102   

                jobtitle  
0              President  
1               VP Sales  
2           VP Marketing  
3   Sales Manager (APAC)  
4    Sale Manager (EMEA)  
5     Sales Manager (NA)  
6              Sales Rep  
7              Sales Rep  
8              Sales Rep  
9              Sales Rep  
10             Sales Rep  
11             Sales Rep  
12             Sales Rep  
13             Sales Rep  
14             Sales Rep  
15             Sales Rep  
16             Sales Rep  
17             Sales Rep  
18             Sales Rep  
19             Sales Rep  
20             Sales Rep  
21             Sales Rep  
22             Sales Rep

There are 23 employees in the company

# 3. What is the total of payments received?

In [4]:
run("""
SELECT SUM(amount)
FROM _payments
""")

sum
0  8853839.23

# 4. List the product lines that contain 'Cars'

In [5]:
run("""
SELECT *
FROM _productlines
WHERE productline = 'Classic Cars' OR productline = 'Vintage Cars'
""")

productline                                    textdescription  \
0  Classic Cars  Attention car enthusiasts: Make your wildest c...   
1  Vintage Cars  Our Vintage Car models realistically portray a...   

  htmldescription image  
0                        
1

In [35]:
run("""
SELECT *
FROM _products
WHERE productline LIKE '%Cars%'
""")

productcode                   productname   productline productscale  \
0     S10_1949      1952 Alpine Renault 1300  Classic Cars         1:10   
1     S10_4757           1972 Alfa Romeo GTA  Classic Cars         1:10   
2     S10_4962        1962 LanciaA Delta 16V  Classic Cars         1:10   
3     S12_1099             1968 Ford Mustang  Classic Cars         1:12   
4     S12_1108             2001 Ferrari Enzo  Classic Cars         1:12   
..         ...                           ...           ...          ...   
57    S24_4258         1936 Chrysler Airflow  Vintage Cars         1:24   
58    S24_4620         1961 Chevrolet Impala  Classic Cars         1:18   
59    S32_4289      1928 Ford Phaeton Deluxe  Vintage Cars         1:32   
60    S50_1341  1930 Buick Marquette Phaeton  Vintage Cars         1:50   
61   S700_2824               1982 Camaro Z28  Classic Cars         1:18   

               productvendor  \
0    Classic Metal Creations   
1    Motor City Art Classics   
2        Second Gear Diecast   
3      Autoart Studio Design   
4        Second Gear Diecast   
..                       ...   
57       Second Gear Diecast   
58   Classic Metal Creations   
59  Highway 66 Mini Classics   
60       Studio M Art Models   
61  Carousel DieCast Legends   

                                   productdescription  quantityinstock  \
0   Turnable front wheels; steering function; deta...             7305   
1   Features include: Turnable front wheels; steer...             3252   
2   Features include: Turnable front wheels; steer...             6791   
3   Hood, doors and trunk all open to reveal highl...               68   
4   Turnable front wheels; steering function; deta...             3619   
..                                                ...              ...   
57  Features opening trunk,  working steering syst...             4710   
58  This 1:18 scale precision die-cast reproductio...             7869   
59  This model features grille-mounted chrome horn...              136   
60   Features opening trunk,  working steering system             7062   
61  Features include opening and closing doors. Co...             6934   

    buyprice    msrp  
0      98.58  214.30  
1      85.68  136.00  
2     103.42  147.74  
3      95.34  194.57  
4      95.59  207.80  
..       ...     ...  
57     57.46   97.39  
58     32.33   80.84  
59     33.02   68.79  
60     27.06   43.64  
61     46.53  101.15  

[62 rows x 9 columns]

# 5. Report total payments for October 28, 2004

In [34]:
run("""
SELECT SUM(amount)
FROM _payments
WHERE paymentdate = '2004-10-28'
""")

sum
0  47411.33

# 6. Report those payments greater than 100000

In [37]:
run("""
SELECT *
FROM _payments
WHERE amount > 100000
""")

customernumber checknumber paymentdate     amount
0             124    AE215433  2005-03-05  101244.59
1             124    KI131716  2003-08-15  111654.40
2             141     ID10962  2004-12-31  116208.40
3             141    JE105477  2005-03-18  120166.58
4             148    KM172879  2003-12-26  105743.00

# 7. List the products in each product line

In [61]:
run("""
SELECT productline, productname
FROM _products
GROUP BY productline, productname
ORDER BY productline ASC
""")

productline                                productname
0    Classic Cars                1948 Porsche 356-A Roadster
1    Classic Cars             1948 Porsche Type 356 Roadster
2    Classic Cars                         1949 Jaguar XK 120
3    Classic Cars                   1952 Alpine Renault 1300
4    Classic Cars                          1952 Citroen-15CV
..            ...                                        ...
105  Vintage Cars  1938 Cadillac V-16 Presidential Limousine
106  Vintage Cars                    1939 Cadillac Limousine
107  Vintage Cars                1939 Chevrolet Deluxe Coupe
108  Vintage Cars                   1940 Ford Delivery Sedan
109  Vintage Cars    1941 Chevrolet Special Deluxe Cabriolet

[110 rows x 2 columns]

In [62]:
run("""
SELECT productline, productname
FROM _products
WHERE productline = 'Classic Cars'
""")

productline                          productname
0   Classic Cars             1952 Alpine Renault 1300
1   Classic Cars                  1972 Alfa Romeo GTA
2   Classic Cars               1962 LanciaA Delta 16V
3   Classic Cars                    1968 Ford Mustang
4   Classic Cars                    2001 Ferrari Enzo
5   Classic Cars                   1969 Corvair Monza
6   Classic Cars                   1968 Dodge Charger
7   Classic Cars                     1969 Ford Falcon
8   Classic Cars              1970 Plymouth Hemi Cuda
9   Classic Cars                   1969 Dodge Charger
10  Classic Cars                      1993 Mazda RX-7
11  Classic Cars                1965 Aston Martin DB5
12  Classic Cars          1948 Porsche 356-A Roadster
13  Classic Cars                     1995 Honda Civic
14  Classic Cars       1998 Chrysler Plymouth Prowler
15  Classic Cars         1999 Indy 500 Monte Carlo SS
16  Classic Cars          1992 Ferrari 360 Spider red
17  Classic Cars                    1985 Toyota Supra
18  Classic Cars                 1969 Dodge Super Bee
19  Classic Cars                1976 Ford Gran Torino
20  Classic Cars       1948 Porsche Type 356 Roadster
21  Classic Cars                1970 Triumph Spitfire
22  Classic Cars            1957 Corvette Convertible
23  Classic Cars                1957 Ford Thunderbird
24  Classic Cars           1970 Chevy Chevelle SS 454
25  Classic Cars                   1970 Dodge Coronet
26  Classic Cars            1966 Shelby Cobra 427 S/C
27  Classic Cars                   1949 Jaguar XK 120
28  Classic Cars  1958 Chevy Corvette Limited Edition
29  Classic Cars                    1952 Citroen-15CV
30  Classic Cars              1982 Lamborghini Diablo
31  Classic Cars            1969 Chevrolet Camaro Z28
32  Classic Cars            1971 Alpine Renault 1600s
33  Classic Cars                  2002 Chevy Corvette
34  Classic Cars              1956 Porsche 356A Coupe
35  Classic Cars    1992 Porsche Cayenne Turbo Silver
36  Classic Cars                1961 Chevrolet Impala
37  Classic Cars                      1982 Camaro Z28

In [63]:
run("""
SELECT productline, productname
FROM _products
WHERE productline = 'Motorcycles'
""")

productline                            productname
0   Motorcycles  1969 Harley Davidson Ultimate Chopper
1   Motorcycles                  1996 Moto Guzzi 1100i
2   Motorcycles   2003 Harley-Davidson Eagle Drag Bike
3   Motorcycles                       2002 Suzuki XREO
4   Motorcycles    1936 Harley Davidson El Knucklehead
5   Motorcycles                       1957 Vespa GS150
6   Motorcycles                      1997 BMW R 1100 S
7   Motorcycles               1960 BSA Gold Star DBD34
8   Motorcycles                1982 Ducati 900 Monster
9   Motorcycles                       1997 BMW F650 ST
10  Motorcycles                      1982 Ducati 996 R
11  Motorcycles              1974 Ducati 350 Mk3 Desmo
12  Motorcycles                     2002 Yamaha YZR M1

In [64]:
run("""
SELECT productline, productname
FROM _products
WHERE productline = 'Planes'
""")

productline                       productname
0       Planes       1980s Black Hawk Helicopter
1       Planes                  Boeing X-32A JSF
2       Planes                    P-51-D Mustang
3       Planes  1928 British Royal Navy Airplane
4       Planes            1900s Vintage Bi-Plane
5       Planes          Corsair F4U ( Bird Cage)
6       Planes           1900s Vintage Tri-Plane
7       Planes       American Airlines: B767-300
8       Planes    America West Airlines B757-200
9       Planes                     ATA: B757-300
10      Planes                F/A 18 Hornet 1/72
11      Planes         American Airlines: MD-11S

In [65]:
run("""
SELECT productline, productname
FROM _products
WHERE productline = 'Ships'
""")

productline                productname
0       Ships     1999 Yamaha Speed Boat
1       Ships      18th century schooner
2       Ships      The Schooner Bluenose
3       Ships              The Mayflower
4       Ships                 HMS Bounty
5       Ships  The USS Constitution Ship
6       Ships                The Titanic
7       Ships             The Queen Mary
8       Ships                 Pont Yacht

In [66]:
run("""
SELECT productline, productname
FROM _products
WHERE productline = 'Trains'
""")

productline                             productname
0      Trains                Collectable Wooden Train
1      Trains  1950's Chicago Surface Lines Streetcar
2      Trains          1962 City of Detroit Streetcar

In [68]:
run("""
SELECT productline, productname
FROM _products
WHERE productline = 'Trucks and Buses'
""")

productline                                  productname
0   Trucks and Buses                               1958 Setra Bus
1   Trucks and Buses                            1957 Chevy Pickup
2   Trucks and Buses                       1940 Ford Pickup Truck
3   Trucks and Buses                       1964 Mercedes Tour Bus
4   Trucks and Buses                        1926 Ford Fire Engine
5   Trucks and Buses                             1940s Ford truck
6   Trucks and Buses                     1962 Volkswagen Microbus
7   Trucks and Buses                  1980’s GM Manhattan Express
8   Trucks and Buses                  1954 Greyhound Scenicruiser
9   Trucks and Buses  1996 Peterbilt 379 Stake Bed with Outrigger
10  Trucks and Buses             Diamond T620 Semi-Skirted Tanker

In [69]:
run("""
SELECT productline, productname
FROM _products
WHERE productline = 'Vintage Cars'
""")

productline                                productname
0   Vintage Cars                       1937 Lincoln Berline
1   Vintage Cars   1936 Mercedes-Benz 500K Special Roadster
2   Vintage Cars                   1917 Grand Touring Sedan
3   Vintage Cars                         1911 Ford Town Car
4   Vintage Cars                  1932 Model A Ford J-Coupe
5   Vintage Cars                     1928 Mercedes-Benz SSK
6   Vintage Cars                1913 Ford Model T Speedster
7   Vintage Cars                         1934 Ford V8 Coupe
8   Vintage Cars        18th Century Vintage Horse Carriage
9   Vintage Cars                          1903 Ford Model A
10  Vintage Cars                   1917 Maxwell Touring Car
11  Vintage Cars    1941 Chevrolet Special Deluxe Cabriolet
12  Vintage Cars        1932 Alfa Romeo 8C2300 Spider Sport
13  Vintage Cars                        1904 Buick Runabout
14  Vintage Cars                    1939 Cadillac Limousine
15  Vintage Cars                1939 Chevrolet Deluxe Coupe
16  Vintage Cars  1938 Cadillac V-16 Presidential Limousine
17  Vintage Cars           1912 Ford Model T Delivery Wagon
18  Vintage Cars                  1937 Horch 930V Limousine
19  Vintage Cars                   1940 Ford Delivery Sedan
20  Vintage Cars           1936 Mercedes Benz 500k Roadster
21  Vintage Cars                      1936 Chrysler Airflow
22  Vintage Cars                   1928 Ford Phaeton Deluxe
23  Vintage Cars               1930 Buick Marquette Phaeton

# 8. How many products in each product line?

SELECT ClientID, ServerID, MAX(`Last.Date`) AS `Last.Date`, COUNT(*) AS ConnectionCount
FROM YourTable
GROUP BY ClientID, ServerID

In [56]:
run("""
SELECT productline, COUNT(productline)
FROM _products
GROUP BY productline
""")

productline  count
0      Classic Cars     38
1            Trains      3
2            Planes     12
3  Trucks and Buses     11
4      Vintage Cars     24
5       Motorcycles     13
6             Ships      9

# 9. What is the minimum payment received?

In [71]:
run("""
SELECT MIN(amount)
FROM _payments
""")

min
0  615.45

The minimum payment received is 615.45

# 10. Which orders have a value greater than 5,000?

In [85]:
run("""
SELECT ordernumber
FROM _orders
WHERE EXISTS (SELECT amount FROM _payments WHERE amount > 5000)
""")

ordernumber
0          10100
1          10101
2          10102
3          10103
4          10104
..           ...
321        10421
322        10422
323        10423
324        10424
325        10425

[326 rows x 1 columns]

In [86]:
run("""
SELECT ordernumber
FROM _orders
WHERE EXISTS (SELECT checknumber FROM _payments WHERE _orders.customernumber = _payments.customernumber AND amount > 5000)
""")

ordernumber
0          10100
1          10101
2          10102
3          10103
4          10104
..           ...
319        10421
320        10422
321        10423
322        10424
323        10425

[324 rows x 1 columns]